### ISSUE TITLE
[receiver/kubeletstat] Review `cpu.utilization` naming

### ISSUE THREAD

**Issue by TylerHelmuth** [open-telemetry/opentelemetry-collector-contrib#27885]
### Component(s)

receiver/kubeletstats

### Is your feature request related to a problem? Please describe.

The Kubeletestats Receiver currently uses `*.cpu.utilization` as the name for cpu metrics that report the [CPUStats](https://pkg.go.dev/k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1#CPUStats) `UsageNanoCores` value. 

I believe that `UsageNanoCores` reports the actual amount of cpu being used not the ratio of the amount being used out of a total limit.  If this is true, then our use of `utilization` is not meeting [semantic convention exceptions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#instrument-naming).

I would like to have a discussion about what exactly `UsageNanoCores` represents and if our metric naming needs updating.

Related to discussion that started in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/24905

**Comment by TylerHelmuth** [comment]
@povilasv thank you!

**Comment by ChrsMark** [comment]
FYI @TylerHelmuth @povilasv In SemConv we have merged https://github.com/open-telemetry/semantic-conventions/pull/282 which adds the `container.cpu.time` metric for now.

For the `dockerstats` receiver we have https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31649 which will try to align the implementation with the added SemConvs.

Do we have a summary so far for what is missing from the `kubeletstats` receiver?

**Comment by TylerHelmuth** [comment]
@ChrsMark in my opinion yes to all questions. We want to be aligned with the spec (although I'd love to reduce the number of iterations in the receivers to gain that alignment, how long till we're stable lol).

I don't have a lot of time to dedicate to getting kubeletstatsreceiver update-to-date with the non-stable spec.  At this point I was planning to wait for things to stabilize before making any more changes besides the work we started in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27885.

**Comment by ChrsMark** [comment]
Thank's @TylerHelmuth, I see the point of not chasing after an unstable schema/spec.

Just to clarify regarding the `container.cpu.utilization` though: shall we abort its deprecation and try to calculate this properly as it was mentioned at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/25901#issuecomment-1776708553? This can happen based on the `cpuNodeLimit`, and seems to be doable based on a quick research I did: https://github.com/ChrsMark/opentelemetry-collector-contrib/commit/27ce7691af67933097ea0defe2bf0f646d19a6ea

**Comment by TylerHelmuth** [comment]
@ChrsMark yes I'd be fine with keeping the metric if we can calculate it correctly.  We'd still need to go through some sort of feature-gate processor to make it clear to users that the metric has changed and that if they want the old value they need to use the new `.usage` metric.

**Comment by ChrsMark** [comment]
@TylerHelmuth @povilasv I have drafted a patch to illustrate the point at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32295. My sightings look promising :). 

If we agree on the idea I can move the PR forward to fix the details and open it for review. Let me know what you think.

**Comment by TylerHelmuth** [comment]
Seems reasonable.  @jinja2 please take a look

**Comment by ChrsMark** [comment]
Thank's for the feedback folks!
I have adjusted https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32295 to use an informer instead of getting the Node on every scrape. Something I noticed is that we have several places where there are different informer based implementations, like in `k8sclusterreceiver` and `k8sattributesprocessor`. Maybe it would make sense to extract that common logic into a common lib and re-use it from the various receivers and processors (even in `resourcedetectorprocessor`), but we can file a different issue for this.
For now we can continue any discussion at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32295.

**Comment by dmitryax** [comment]
Hey folks. Sorry, I'm late to the party.

I am not convinced about repurposing the `container.cpu.utilization` to be a ratio of the node cpu limit:
1. It's unclear from the metric's name what the limit, while `k8s.container.cpu_limit_utilization` gives pretty good idea. I can imagine it to be utilization of some container limit, not node's limit. 
2. `container.cpu.utilization` name is generic and can be used in docker receiver as well. Can we always easily retrieve number of host CPUs there? I'm not sure about it.

I think something like `k8s.container.node_limit_utilization` would be a better name here. It's clear and consistent with other `utilization` metrics.

Even if we decide to repurpose `container.cpu.utilization`, I would strongly suggest deprecating->disabling and maybe removing it first instead of just changing its meaning. I believe the metrics based on the node limit should be optional, and `container.cpu.time` or `container.cpu.usage` should be enabled by default instead.

cc @ChrsMark @TylerHelmuth @povilasv

**Comment by ChrsMark** [comment]
Thank's @dmitryax. I agree with making the metric more specific here.
There is a similar distinction in [meatricbeat](https://github.com/elastic/beats/blob/ffcd1814666645a5d7a644911ecf6e2b7d8db3f5/metricbeat/module/kubernetes/container/_meta/fields.yml#L30-L39) as well, with `*.node.pct` and `*.limit.pct` accordingly.

If others agree, I can change the PR accordingly to:
1. introduce the `k8s.container.cpu.node_limit_utilization`.
2. remove the feature flag since we don't need it any more and leave `container.cpu.utilization` as deprecated.
3. since we are on it, I suggest we also introduce the `k8s.pod.cpu.node_limit_utilization` as part of the same PR.
4. for `k8s.node.cpu.utilization` since we can use it as is I guess we will need to go through the feature flag path in a separate PR.

Extra: I wonder also if that would make sense to change `k8s.container.cpu_limit_utilization` to `k8s.container.cpu.limit_utilization` so as to be consistent with having `k8s.container.cpu.*` as a clear namespace.

@dmitryax @TylerHelmuth @povilasv let me know what you think.

**Comment by jinja2** [comment]
What does "node_limit" mean here? Is it referring to the host's capacity or the [node allocatable](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable) (capacity - system/kube reserved)? I find node_limit to be ambiguous but my assumption would be that it refers to allocatable since that's the amount of resources actually available for pod scheduling. Do you think a user might want to select whether the utilization is against the capacity or the allocatable?

**Comment by TylerHelmuth** [comment]
> Even if we decide to repurpose container.cpu.utilization, I would strongly suggest deprecating->disabling and maybe removing it first instead of just changing its meaning. I believe the metrics based on the node limit should be optional, and container.cpu.time or container.cpu.usage should be enabled by default instead.

I am good with this.  I care mainly about the switch from `*.cpu.utilization` to `*.cpu.usage`.

**Comment by ChrsMark** [comment]
> Do you think a user might want to select whether the utilization is against the capacity or the allocatable?

I would go with the `capacity` since the `allocatable` is kind of k8s specific and is another "limit" to my mind. I think users would be fine to see the actual utilization against the actual node's capacity. This would not require extra knowledge like the allocatable configuration etc.
__________________

I will change the [PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32295) for now to to use the `k8s.container.cpu.node_limit_utilization` and we can continue the discussion there.

**Comment by ChrsMark** [comment]
Any thoughts about the [k8s.node.cpu.utilization](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.104.0/receiver/kubeletstatsreceiver/metadata.yaml#L92-L100)? Should this be calculated properly instead of deprecated? This one should be straightforward compared to the container's/pod's ones since the node does not have the respective request/limit based metrics.

So it would just be
```
k8s.node.cpu.utilization = k8s.node.cpu.usage / num_of_cores
```

**Comment by ChrsMark** [comment]
I wonder though if we should first deprecate the metrics and re introduce this specific one back as it was pointed out at https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27885#issuecomment-2089636637.

Otherwise we can fix this one directly.

**Comment by ChrsMark** [comment]
The thing is that change the way it is calculated requires enabling the `NodeInformer`. So if we change this now as is, any existing users' configurations will become invalid and the metric will being reported (will be `0`). This is highlighted by the failing tests at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34191.

So what I would suggest here is:
1) Make the `*.cpu.usage` metrics enabled by default.
2) Make the `*.cpu.utilization` metrics disabled by default.
3) Remove entirely the `k8s.node.cpu.utilization`, `k8s.pod.cpu.utilization` and `container.cpu.utilization` metrics.
4) Add back the `k8s.node.cpu.utilization` with https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34191, which will be disabled by default.

The above should be done in concrete/standalone PRs and can be split across different releases to ensure a gradual switch process for the users.

WDYT?

/cc @TylerHelmuth @dmitryax

**Comment by TylerHelmuth** [comment]
> So if we change this now as is, any existing users' configurations will become invalid and the metric will being reported (will be 0).

Ya that'd be bad.

I like this plan as it works toward the original goal (using the proper name) and allows us to work on an actual utilization metric separately. My only concern is that I really didn't want to do breaking semantic convention changes to the k8s components until the k8s semconv was stable, so that all the changes could be hidden behind one feature flag.

That isn't happening anytime soon tho. I think starting with enabling `cpu.usage` by default is a good next step.

**Comment by dmitryax** [comment]
@ChrsMark, I like that outlined plan. Thanks for putting it together! However, I'd prefer combining 1 and 2 in one release so we don't change the cardinality.

**Comment by TylerHelmuth** [comment]
@dmitryax If we do 1 and 2 in one release we need more evangelism around the change. The `usage` metrics and their warnings have been around for a long time, but the change wasn't very public and I worry users aren't adopting it.

I think we need to update to the README, add a warning in the release notes, and update the printed warning that the value will "swap" in a specific future release. Maybe write a blog post too.

**Comment by ChrsMark** [comment]
> I think we'd need to update to the README, add a warning in the release notes, and update the printed warning that the value will "swap" in a specific future release. Maybe write a blog post too.

@TylerHelmuth @dmitryax Should we then first do those in a step 0 in order to later combine step 1 and step 2? I'd be fine with that and taking care of it.

**Comment by TylerHelmuth** [comment]
Sure

**Comment by ChrsMark** [comment]
@TylerHelmuth cool! Any preferences on what the target release of this change/deprecation should be? Based on the [release-schedule](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/release.md#release-schedule)  I would go with `v0.111.0` (`2024-10-07`).

**Comment by TylerHelmuth** [comment]
Is there a way for use to add a feature gate or something that causes the collector to fail on start if `utilization` is used? Breaking changes from semantics are so difficult because a user could upgrade the collector and not realize it just started emitting new telemetry and in this case it is a metric that has been a core metric from this receiver for 4+ years.

It would be great it we can add a set where the collector fails to start if utilization is in use, and the user could disable the feature gate to get back to a running state, but at least they'd have to acknowledge the change is coming on a specific date.

**Comment by ChrsMark** [comment]
> It would be great it we can add a set where the collector fails to start if utilization is in use, and the user could disable the feature gate to get back to a running state, but at least they'd have to acknowledge the change is coming on a specific date.

Hmm having the Collector to fail completely for one metric sounds a bit concerning to me tbh. Has this approach been used in other components too?

How about making the switch completely behind a feature gate then?

This way we enable users to switch whenever they want prior to a specific release and then after we have make the switch the feature gate to continue using the `utilization` will still be there for some period. We can define a plan like the following:

- Introduce it in v0.x.0 as `alpha` - disabled by default.
- Move it to `beta` in v0.y.0 - enabled by default. (for configs that still enable the `utilization` it can fail)
- Move it to `stable` in v0.z.0  - cannot be disabled.
- Remove it three releases after `stable`.

(we have something similar in `hostmetricsreceiver`)

Wouldn't that cover most of what we have discussed here?

**Comment by TylerHelmuth** [comment]
Yes thats what I'm thinking.  If the feature gate is enabled and a `utilization` metric is enabled, the collector can fail.

**Comment by ChrsMark** [comment]
Thank's @TylerHelmuth! I will be off for the following weeks, but I will pick it up once I'm back.

**Comment by ChrsMark** [comment]
@TylerHelmuth @dmitryax I have tried the feature gate approach at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35139. Please have a look once you get the chance. You can find more details about the feature gate's plan in the PR's description.

**Comment by ChrsMark** [comment]
@TylerHelmuth @dmitryax based on the plan discussed at https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27885#issuecomment-2285476787, it's good timing now to move the feature gate to beta - enabled by default.

The feature gate was introduced by https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35139 in `v0.111.0` so it's 10+ releases in.

Let me know if there is any objection, otherwise I will send a PR soon.

**Comment by TylerHelmuth** [comment]
No objections

**Comment by dmitryax** [comment]
Now after moving from `*.utilization` to `*.usage`, the kubeletstats receiver still has 2 metrics enabled by default reporting the same data in different forms. This should be probably addressed once we align on https://github.com/open-telemetry/semantic-conventions/issues/2177

**Comment by Erog38** [comment]
> The thing is that change the way it is calculated requires enabling the `NodeInformer`. So if we change this now as is, any existing users' configurations will become invalid and the metric will being reported as `0`. This is highlighted by the failing tests at [#34191](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34191).
> 
> So what I would suggest here is:
> 
> 0. Update the README file, add a warning in the release notes, and update the printed warning that the value will "swap" in a specific future release. Maybe write a blog post too.
> 1. Make the `*.cpu.usage` metrics enabled by default: [[receiver/kubeletstats] Enable by default the .cpu.usage metrics #34217](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34217)
> 2. Make the `*.cpu.utilization` metrics disabled by default.
> 3. Remove entirely the `k8s.node.cpu.utilization`, `k8s.pod.cpu.utilization` and `container.cpu.utilization` metrics.
> 4. Add back the `k8s.node.cpu.utilization` with [[receiver/kubeletstats] Emit k8s.node.cpu.utilization as ratio against node's capacity #34191](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34191), which will be disabled by default.
> 
> The above should be done in concrete/standalone PRs and can be split across different releases to ensure a gradual switch process for the users.
> 
> WDYT?


If this is still the plan, is there a stopgap in place to still get cpu % out of 100 for a k8s node? As of right now, `k8s.node.cpu.utilization` is gone, and [`k8s.node.allocatable.cpu`](https://opentelemetry.io/docs/specs/semconv/system/k8s-metrics/#metric-k8snodeallocatablecpu) doesn't exist, so we can't calculate this value.

How far are we from re-introducing this metric as properly computed?

**Comment by ChrsMark** [comment]
> How far are we from re-introducing this metric as properly computed?

@Erog38 first we need to complete the removal of the deprecated metrics with https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/41178. There is an ongoing PR for this. Then we will need a follow-up issue to discuss re-introducing the metric as properly calculated. 

BTW, there is already a node allocatable metric from the k8scluster receiver that you can use: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/51156971cca93c639a9d3113178c5692e7699ec1/receiver/k8sclusterreceiver/internal/node/nodes.go#L220. You need to enable those with the [`allocatable_types_to_report`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/k8sclusterreceiver/README.md?plain=1#L56) setting. (probably the docs need to explicitly mention the metrics produced by this setting to avoid confusion. 

Note that [k8s.node.allocatable.cpu](https://opentelemetry.io/docs/specs/semconv/system/k8s-metrics/#metric-k8snodeallocatablecpu) is not ported back to the receiver from SemConv: https://opentelemetry.io/docs/specs/semconv/non-normative/k8s-migration/#k8s-node-allocatable-metrics

**Comment by Erog38** [comment]
@ChrsMark Thank you so much for the update and context! I'll get the allocatable metric from the k8s_cluster receiver for now. I believe that should work for our use cases at the moment. I went looking for it, but I wasn't aware of the rename there as well. 

If you end up with an orphaned issue to help this along and need help, feel free to @ me and I'll help however I can.

**Comment by ChrsMark** [comment]
Sounds good @Erog38! This issue will be updated once we have progress so you will get notified.

**Comment by ChrsMark** [comment]
After the merge of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/42727 the `k8s.node.cpu.utilization`, `k8s.pod.cpu.utilization` and `container.cpu.utilization` metrics have been deprecated and removed  completely. In this, we can close this issue now.

**Comment by shanehull** [comment]
> 4. Add back the k8s.node.cpu.utilization with https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34191, which will be disabled by default.

Has this step been abandoned? It would be nice to have a cpu utilization metric as a percent of the node's capacity.

**Comment by ChrsMark** [comment]
@shanehull this will need to go through https://github.com/open-telemetry/semantic-conventions/issues/2768 first.
